pkg/sql/exec_log.go

```diff
@@ -92,13 +93,13 @@ var logStatementsExecuteEnabled = settings.RegisterBoolSetting(
 
 // maybeLogStatement conditionally records the current statement
 // (p.curPlan) to the exec / audit logs.
-func (p *planner) maybeLogStatement(ctx context.Context, lbl string, rows int, err error) {
+func (p *planner) maybeLogStatement(ctx context.Context, lbl string, ast tree.Statement, rows int, err error) {
```

Comment:
Wait why did this need to change?

This changes gives the impression it is valid to call maybeLogStatement with an AST that's unrelated to the state currently collected in the planner (the current `PhaseTimes`). This would be incorrect. The interface for this function was defined that way precisely to prevent giving that impression.

Reply:
Prior we would extract it from `planner.curPlan`, however, assigning the AST to the plan *previously* stored on the planner is clunky and repetitive, which we would need to do in situations where the optimizer doesn't produce a plan (specifically, error cases).

I think I disagree about the interface here - in the presence of the optimizer, where the `planner` struct isn't doing the planning work, use of `curPlan` for threading state around is error prone. It's very easy to assign something to that plan only to get it overwritten later by the plan generated by the optimizer (this is the second such issue I've run into, the first being `auditEvents`).

My preference would be to either leave this change as it is, or to lift the `AST` field onto the planner, rather than putting it on `planner.curPlan`.

Reply:
Your notion that the `curPlan` field is reserved to the heuristic planner is erroneous. 

The `curPlan` field does what it says on the label: it should contain whatever state is needed for the plan currently being executed, whichever planning code was used to produce it.

I would recommend that the AST be stored in `curPlan` upfront regardless of which planning code is going to be used.

Reply:
the `execFactory` produces an entirely new plan, and we do store the AST on the resulting plan after it's been successfully constructed. This particular problem is concerned with what happens when we *fail to construct a plan*, in which case I feel it's justified to avoid depending on values which live on `curPlan`, unless we want to extend the semantics of `curPlan` to also contain certain data in the case of failure, which feels iffy to me.

Reply:
The error you are making is that `makeOptimizerPlan` should populate `curPlan.AST` regardless of whether an error is encountered.

This behavior should be symmetric with `makePlan` (all the more since it intends to replace it, later):

```
func (p *planner) makePlan(ctx context.Context, stmt Statement) error {
        // Reinitialize.
        p.curPlan = planTop{AST: stmt.AST}

        var err error
        p.curPlan.plan, err = p.newPlan(ctx, stmt.AST, nil /*desiredTypes*/)
        if err != nil {
                return err
        }
        // ...
```

So if you were to move the assignment `        p.curPlan.AST = stmt.AST` in `makeOptimizerPlan` to be at the top like in `makePlan`, we'd be golden.

Reply:
We can't just move the assignment without threading a bunch of state through the `execFactory`, but we can put it twice. I still think this approach will be error-prone, but I don't care enough to debate it.

Reply:
It is either that or changing the various other interfaces to thread the `planTop` as argument and remove it from the planner entirely (this might be a good idea). 

Having half of the state carried implicitly via the planner and the other half in arguments is the error-prone approach.